spyder/api/widgets.py

```diff
@@ -0,0 +1,1478 @@
+
+        return actions
+
+    def add_corner_widget(self, name, widget, before=None):
+        """
+        Add widget to corner, that is to the left of the last added widget.
+
+        Parameters
+        ----------
+        name: str
+            Unique name of the widget.
+        widget: QWidget
+            Any QWidget to add in the corner widget.
+        before: QWidget
+            Insert the widget before this widget.
+
+        Notes
+        -----
+        By default widgets are added to the left of the last corner widget.
+
+        SpyderWidget provides an options menu button and a spinner so any
+        additional widgets will be placed to the left of the spinner,
+        if visible.
+        """
+        if self._corner_widget is not None:
+            self._corner_widget.add_widget(name, widget)
+
+    def get_corner_widget(self, name):
+        """
+        Return the a widget inside the corner widget by name.
+
+        Parameters
+        ----------
+        name: str
+            Unique name of the widget.
+        """
+        return self._corner_widget.get_widget(name)
+
+    def start_spinner(self):
+        """
+        Start default status spinner.
+        """
+        self._spinner.setVisible(True)
+        self._spinner.start()
+
+    def stop_spinner(self):
+        """
+        Stop default status spinner.
+        """
+        self._spinner.stop()
+        self._spinner.setVisible(False)
+
+    def create_toolbar(self, name, location='top'):
+        """
+        Create and add an auxiliary toolbar at the top or at the bottom
+        of the plugin.
+
+        Parameters
+        ----------
+        location: str
+            A string whose value is used to determine where to add the
+            toolbar in the plugin interface. The toolbar can be added either
+            at the 'top' or at the 'bottom' of the plugin.
+
+        Returns
+        -------
+        SpyderPluginToolbar
+            The auxiliary toolbar that was created and added to the plugin
+            interface.
+        """
+        if name in self._toolbars:
+            raise Exception('!')
+
+        if location not in ['top', 'bottom']:
+            raise Exception('Invalid location "{}"!'.format(location))
+
+        toolbar = MainWidgetToolbar(parent=self)
+        self._toolbars[name] = toolbar
+        self._add_auxiliary_toolbar(toolbar, location)
+
+        return toolbar
+
+    def create_menu(self, name, text=''):
+        """
+        Override SpyderMenuMixin method to use a different menu class.
+        """
+        menus = getattr(self, '_menus', None)
+        if menus is None:
+            self._menus = OrderedDict()
+
+        if name in self._menus:
+            raise Exception('Menu name "{}" already in use!'.format(name))
+
+        menu = MainWidgetMenu(name, title=text, parent=self)
+        self._menus[name] = menu
+        return menu
+
+    def _add_auxiliary_toolbar(self, toolbar, location):
```

Comment:
When is this necessary?

Reply:
Ideally a panel should only have a single set of tool buttons or widgets on top, like the variable explorer, the plots pane etc.

Some plugins are a bit "weird" and verbose and add "two lines" (like the pylint one) or 3 lines "find in files" 

These are now all ToolBars, so anythong beyondthe main one, is an auxiliary one.

Reply:
On this subject, part of reviewing the UI once this migration occurs is to take some decisions on what we want and enforce that in an uniform way for all plugins.

Some things to take into account:
- [ ] Names vs no names next to toolbuttons (I would say, no names for all!)
- [ ] Simplification of plugins with too many buttons, options or "lines of widgets"on the top of the pane (like find in files...)
- [ ] Remove duplicate buttons like Start/Stop, and replace by a single one (like find in files now does)